File: Data Structures & Algorithms/three-integer-sum/submission-0.java
class Solution {
    // T = O(n^2) where n is the number of elements in nums
    // AUX = O(log n)
    public List<List<Integer>> threeSum(int[] nums) {
        // AUX = O(n) where n is the number of elements in nums
        Set<List<Integer>> triplets = new HashSet();
        Arrays.sort(nums); // T = (n log n), AUX = (log n), where n is the number of elements in nums

        // T = O(n^2)
        for (int i = 0; i < nums.length; i++) {
            int l = i + 1, r = nums.length - 1;
            // T = O(n)
            while (l < r) {
                int sum = nums[i] + nums[l] + nums[r];

                if (sum == 0) {
                    triplets.add(List.of(nums[i], nums[l], nums[r]));
                    l++;
                    r--;
                } else if (sum > 0) {
                    r--;
                } else {
                    l++;
                }
            }
        }

        // Space complexity disregarded because it is used for the result
        List<List<Integer>> result = new ArrayList();

        // O(n)
        for (List<Integer> triplet : triplets) {
            result.add(triplet);
        }

        return result;
    }
}

File: Data Structures & Algorithms/concatenation-of-array/submission-0.py
class Solution:
    # nums = [1, 2, 1, 4], l = 4
    def getConcatenation(self, nums: List[int]) -> List[int]:
        # [0, 0, 0, 0, 0, 0, 0, 0]
        ans = [0 for i in range(len(nums) * 2)]
        i = 0

        for j in range(len(nums), len(ans)):
            ans[i] = nums[i]
            ans[j] = nums[i]
            i += 1
        
        return ans